backend/checkmach.ml

```diff
@@ -27,454 +27,809 @@
-      (* conservative *)
-      let raise_after = check_instr_exn t i.next raise_after in
-      let _ = check_instr_exn t body raise_after in
-      List.iter
-        (fun (_n, _ts, handler) -> check_instr_exn t handler false |> ignore)
-        handlers;
-      false
-    | Itrywith (body, _kind, (_ts, handler)) ->
-      (* conservative *)
-      let _ = check_instr_exn t i.next raise_after in
-      let _ = check_instr_exn t body false in
-      let _ = check_instr_exn t handler false in
-      false
-    | Iraise _ -> false
-    | Ireturn _ -> false
-    | Iexit _ -> false
-
-  let debug t expected =
-    match Unit_info.get_value unit_info t.fun_name with
-    | None -> assert false
-    | Some v ->
-      (* just checking for consistency with the recorded value *)
-      if not (v = expected)
-      then
-        Misc.fatal_errorf
-          "Checkmach %s: Wrong result for %s, expected %s, found %s" S.name
-          t.fun_name (Value.to_string expected) (Value.to_string v)
-
-  let fundecl ppf (f : Mach.fundecl) =
+      let r = S.transform_specific s ~next ~exn in
+      check t r "Arch.specific_operation" dbg
+
+  module D = Dataflow.Backward ((Value : Dataflow.DOMAIN))
+
+  let check_instr t body =
+    let transfer (i : Mach.instruction) ~next ~exn =
+      match i.desc with
+      | Ireturn _ -> Value.normal_return
+      | Iop op -> transform_operation t op ~next ~exn i.dbg
+      | Iraise Raise_notrace ->
+        (* [raise_notrace] is typically used for control flow, not for
+           indicating an error. Therefore, we do not ignore any allocation on
+           paths to it. The following conservatively assumes that normal and exn
+           Returns are reachable. *)
+        Value.join exn Value.safe
+      | Iraise (Raise_reraise | Raise_regular) -> exn
+      | Iend -> next
+      | Iexit _ ->
+        report t next ~msg:"transform" ~desc:"iexit" i.dbg;
+        next
+      | Iifthenelse _ | Iswitch _ -> next
+      | Icatch (_rc, _ts, _, _body) ->
+        report t next ~msg:"transform" ~desc:"catch" i.dbg;
+        next
+      | Itrywith (_body, (Regular | Delayed _), (_trap_stack, _handler)) ->
+        report t next ~msg:"transform" ~desc:"try-with" i.dbg;
+        next
+    in
+    (* By default, backward analysis does not check the property on paths that
+       diverge (non-terminating loops that do not reach normal or exceptional
+       return). All loops must go through an (Iexit label) instruction or a
+       recursive function call. If (Iexit label) is not backward reachable from
+       the function's Normal or Exceptional Return, either the loop diverges or
+       the Iexit instruction is not reachable from function entry.
+
+       To check divergent loops, the initial value of "div" component of all
+       Iexit labels is set to "Safe" instead of "Bot". This is conservative with
+       respect to non-recursive Icatch. *)
+    D.analyze ~exnescape:Value.exn_escape ~init_lbl:Value.diverges ~transfer
```

Comment:
`init_lbl` will set `div` to `Safe` for all labels,
including `Iexit` ones, right?

Reply:
Yes, including `Iexit` labels used by non-recursive Icatch handlers.

Reply:
I may have misunderstoond your question. Are you asking about labels used for `Itrywith` handlers rather than `Icatch` handlers? I think with `flambda2` these will also get `div` set to `Safe` instead of `Bot` by default, unfortunately, in the current version.

Reply:
Updated comment.